### DISCUSSION TITLE
Custom event_handler exceptions

### DISCUSSION THREAD

**Discussion by royassis** [aws-powertools/powertools-lambda-python#2221] [Q&A] [answered]
Hey

I want to create a custom exception class when using the **APIGatewayRestResolver** event_handler, is this available ? 

I'll describe my use case: 
I'm triggering an async step-functions job using a lambda, I want to return a 4xx error when the same lambda is triggered while 
the step-function is running. 
Right not im returning a 500 error code, with the **InternalServerError** exception class, but it doesn't feel right as it's not an internal server error. 

Best,
Roy

**Comment by leandrodamascena** [comment, marked as answer]
Hi @royassis! Thanks for starting this thread.

The short answer is yes, you can do that by extending the `ServiceError` class and creating your own exception class, but you don't have to. You already can do this by raising an exception using the `ServiceError` class and informing the HTTP status code and message you want. Let me give you an example of both situations:

**Using the ServiceError class.**
```python
from aws_lambda_powertools import Logger
from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.event_handler.exceptions import (
    ServiceError,
)
from aws_lambda_powertools.logging import correlation_paths
from aws_lambda_powertools.utilities.typing import LambdaContext

logger = Logger()
app = APIGatewayRestResolver()

@app.get(rule="/already-running-route")
def conflict_error():
    raise ServiceError(409, "CONFLICT -> Already running")

@logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_REST)
def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```

**Creating your own exception class**
```python
from http import HTTPStatus

from aws_lambda_powertools import Logger
from aws_lambda_powertools.event_handler import APIGatewayRestResolver
from aws_lambda_powertools.event_handler.exceptions import (
    ServiceError,
)
from aws_lambda_powertools.logging import correlation_paths
from aws_lambda_powertools.utilities.typing import LambdaContext

logger = Logger()
app = APIGatewayRestResolver()

class ConflictError(ServiceError):
    """HTTP 409 conflict error class"""

    def __init__(self, message: str):
        super().__init__(HTTPStatus.CONFLICT, message)


@app.get(rule="/already-running-route")
def conflict_error():
    raise ConflictError("CONFLICT -> Already running")

@logger.inject_lambda_context(correlation_id_path=correlation_paths.API_GATEWAY_REST)
def lambda_handler(event: dict, context: LambdaContext) -> dict:
    return app.resolve(event, context)
```

The result for both examples is
![image](https://user-images.githubusercontent.com/4295173/236703494-f5a72959-b948-40f6-9baa-27aeab1e119b.png)

I used `HTTP Status Code` 409 because I think it's the best fit for this situation, but you can use whatever you like.

Please let me know if you still have any questions about this and I can help you.
Thank you
